rfcs/031-relation_batcher/README.md

```diff
@@ -0,0 +1,112 @@
+
+The main issue is that for a single input to the `relation_embedder`, it can result in many works being denormalised. Given the following tree:
+
+```
+A
+|-----
+|    |
+B    C
+|    |-----
+|    |    |
+D    E    F
+```
+
+When a message is received for work `D` it will need denormalise 2 works, namely `B` and `D`: it needs to denormalise the work itself, and it's parent `B` due to it needing to update it's `children` field. When a message is received for work `A` however it will need to denormalise every work in the tree, due to all of the works containing `A` as an ancestor.
+
+When building the `relation_embedder` the assumption was made that many more works will be further down the tree and leaf nodes than being near the root, so the amount of work performed will be manageable. However this assumption doesn't really hold, with a combinatorially large amount of work being generated by nodes that are high in the tree of large archives.
+
+There are also extreme cases where even though most of the works in an archive are leaf nodes, the archive is very flat. To take one real world example:
+
+```
+MS9225
+|
+|---
+|  |
+1  2
+   |-------------
+   |  |  |      |
+   1  2  3 ... 3583
+```
+
+There are 3583 works here which are siblings of each other, and when any one of these is received at the input it will result in all of their siblings and the two ancestors needing to be denormalised too. For an initial reindex that will result in denormalising 6,427,905 works (`1 + 2 + ... + 3585`), and in a reharvest where in the worst case all of the works in the archive are sent again it will result in 12,852,225 (`3585 * 3585`). In an ideal scenario, we would of course only denormalise each of the works in that archive one time only (i.e. 3,585 denormalisations), with anything above that being unnecessary work. Note this is just a single archive, and there will likely be other similar cases out of the 6349 archives currently in Calm.
+
+## Proposed Solution
+
+In order to prevent the large amount of duplicate work performed within the `relation_embedder`, we propose adding two more small services to the pipeline before the `relation_embedder`: a `router` which extracts archive works and passes through non-archive works, and a `batcher` which groups works contained within the same archives over a given time period, preventing the `relation_embedder` from performing the excessive levels of duplicate work we are currently seeing.
+
+### Router
+
+The `router` takes a work id, looks it up in the `pipeline_storage` and checks if it's part of an 
+archive (i.e. has a `collectionPath` field).  If it is, it sends the value of `collectionPath` to the `batcher`, otherwise sending the work straight to the `id_minter_works`. 
+
+### Batcher
+
+The `batcher` groups work received over a given time period and works out the optimal set of paths to send towards 
```

Comment:
How is the number of works here bounded?

Reply:
In practice we will (as I mentioned above) use [`groupedWithin`](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/groupedWithin.html) and set a very large number (e.g. 100,000).

Reply:
The messages consist just of paths. Say we have 100,000 paths which are 50 characters long, that is only 5MB, so memory is not going to be an issue with number of these sorts of sizes.